README.md
README.md
================

# seedr

![](seeds.png)

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/efernandezpascual/seedr.svg?branch=master)](https://travis-ci.com/efernandezpascual/seedr)
<!-- badges: end -->

## Overview

`seedr` is an R package that provides functions to fit hydro and thermal
time germination models. These models characterize seed lots by two sets
of parameters: (i) the physiological thresholds (water, temperature)
between which the seed lot can germinate, and (ii) the
physiological-time units that the seed lot needs to accumulate before it
can germinate. `seedr` fits the hydro time model of Bradford (Gummerson
1986, Bradford 1990, Bewley et al. 2013) and the thermal time model of
Garcia-Huidobro (Garcia-Huidobro et al. 1982, Gummerson 1986, Bewley et
al. 2013). It allows to fit models to grouped datasets, i.e. datasets
containing multiple species, seedlots or experiments.

## Installation

``` r
# Install last version from GitHub
install.packages("devtools") # Install devtools if you don't have it
devtools::install_github("efernandezpascual/seedr") # Get seedr from GitHub
```

## Usage

``` r
library(seedr)
# seedr relies on physiotime(), a wrapper function that formats your data and fits a hydro/thermal time model
physiotime(grasses, # germination dataset
           x = "psi", # experimental treatment
           method = "bradford", # hydrotime model
           groups = c("species")) # group dataset by species
```

    ## A list of physiological time germination models 
    ##  calculated for the following 2 groups: 
    ##  
    ## species: Anisantha rubens 
    ## Bradford's hydrotime model 
    ##  Water potential levels in experiment: -1.6 -1 -0.8 -0.6 -0.4 -0.2 0 
    ##  Theta - Hydrotime constant: 50.54 
    ##  Psib50 - Base water potential (median): -1.59 
    ##  Sigma of the base water potential: 0.4 
    ##  R2: 0.94 
    ##  
    ## species: Bromus hordeaceus 
    ## Bradford's hydrotime model 
    ##  Water potential levels in experiment: -1.6 -1 -0.8 -0.6 -0.4 -0.2 0 
    ##  Theta - Hydrotime constant: 59.81 
    ##  Psib50 - Base water potential (median): -1.75 
    ##  Sigma of the base water potential: 0.38 
    ##  R2: 0.86 
    ## 

``` r
# for more information read the seedr vignette
vignette("seedr")
```

## References

  - Bewley, J. D., Bradford, K. J., Hilhorst, H. W., & Nonogaki, H.
    (2013). Environmental Control of Germination. In *Seeds: Physiology
    of Development, Germination and Dormancy, 3rd Edition*
    (pp. 302-317). Springer, New York, NY.
  - Bradford, K. J. (1990). A water relations analysis of seed
    germination rates. *Plant Physiology*, 94(2), 840-849.
  - Garcia-Huidobro, J., Monteith, J. L., & Squire, G. R. (1982). Time,
    temperature and germination of pearl millet (*Pennisetum typhoides*
    S. & H.) I. Constant temperature. *Journal of Experimental Botany*,
    33(2), 288-296.
  - Gummerson, R. J. (1986). The effect of constant temperatures and
    osmotic potentials on the germination of sugar beet. *Journal of
    Experimental Botany*, 37(6), 729-741.


NEWS.md
## version 0.3.0

This is the first public release.


R/seedr.R
#' seedr: Hydro and Thermal Time Seed Germination Models in R
#'
#' The \code{seedr} package provides functions to fit hydro and thermal time
#' germination models. These models characterize seed lots by two sets of
#' parameters: (i) the physiological thresholds (water, temperature) between
#' which the seed lot can germinate, and (ii) the physiological-time units that
#' the seed lot needs to accumulate before it can germinate. \code{seedr} allows
#' to fit the hydro time model of Bradford (Gummerson 1986, Bradford 1990,
#' Bewley et al. 2013) and the thermal time model of Garcia-Huidobro
#' (Garcia-Huidobro et al. 1982, Gummerson 1986, Bewley et al. 2013).
#' \code{seedr} also allows to quickly fit models to multi-seedlot or
#' multi-species datasets.
#' @references Bewley, J. D., Bradford, K. J., Hilhorst, H. W., & Nonogaki, H.
#'   (2013). Environmental Control of Germination. In Seeds: Physiology of
#'   Development, Germination and Dormancy, 3rd Edition (pp. 302-317). Springer,
#'   New York, NY.
#'
#'   Bradford, K. J. (1990). A water relations analysis of seed germination
#'   rates. Plant Physiology, 94(2), 840-849.
#'
#'   Garcia-Huidobro, J., Monteith, J. L., & Squire, G. R. (1982). Time,
#'   temperature and germination of pearl millet (Pennisetum typhoides S. & H.)
#'   I. Constant temperature. Journal of Experimental Botany, 33(2), 288-296.
#'
#'   Gummerson, R. J. (1986). The effect of constant temperatures and osmotic
#'   potentials on the germination of sugar beet. Journal of Experimental
#'   Botany, 37(6), 729-741.
#' @import data.table
#' @import graphics
#' @importFrom grDevices colorRampPalette
#' @importFrom stats lm qnorm cov var
#' @docType package
#' @name seedr
NULL

#' Water potential example dataset
#'
#' This is a dataset containing information from a water potential experiment
#' with grass seeds. It is used to give examples of the functions dealing with
#' hydrotime germination models. It also gives and idea of the format in which
#' germination data should be provided to seedr.
#'
#' @format A data frame with 1605 rows and 7 variables \describe{
#'   \item{species}{Name or code for the species to which the data refers}
#'   \item{temperature}{Temperature treatment (in ºC) of the experiment}
#'   \item{psi}{Water potential treatment (in MPa) of the experiment}
#'   \item{dish}{Code for the Petri dish, container, replicate, etc.}
#'   \item{times}{Time of germination scoring since the start of the experiment,
#'   in days} \item{germinated}{Number of germinated seeds recorded at that
#'   time} \item{germinable}{Total number of viable seeds in the replicate} }
#' @source Own data from a laboratory experiment.
"grasses"

#' Temperature example dataset
#'
#' This is a dataset containing information from a germination temperature experiment
#' with centaury seeds. It is used to give examples of the functions dealing with
#' thermal time germination models. It also gives and idea of the format in which
#' germination data should be provided to seedr.
#'
#' @format A data frame with 896 rows and 7 variables \describe{
#'   \item{species}{Name or code for the species to which the data refers}
#'   \item{population}{Name or code for the seedlot}
#'   \item{temperature}{Temperature treatment (in ºC) of the experiment}
#'   \item{dish}{Code for the Petri dish, container, replicate, etc.}
#'   \item{times}{Time of germination scoring since the start of the experiment,
#'   in days} \item{germinated}{Number of germinated seeds recorded at that
#'   time} \item{germinable}{Total number of viable seeds in the replicate} }
#' @source Own data from a laboratory experiment.
"centaury"


R/huidobro.R
#' Fits Garcia-Huidobro's thermal time model
#'
#' \code{huidobro} fits a thermal time seed germination model using the method
#' of Garcia-Huidobro (Garcia-Huidobro et al. 1982, Gummerson 1986, Bewley et
#' al. 2013). This function can be used only with one-group dataset, i.e. one
#' seed lot of one species. To fit models to grouped datasets (multi-seedlots,
#' multi-species) use the function \code{physiotime} instead.
#'
#' @usage huidobro(d, min.ptos = 3, tops = c("Max R2","Max value"), fractions =
#'   (1:9)/10)
#' @param d a data.table within a "physiodata" object, containing the cumulative
#'   germination proportion at each scoring time and temperature treatment.
#' @param min.ptos minimal number of data points (i.e. different temperature
#'   treatments) needed to fit the suboptimal and supraoptimal germination
#'   models. If the number of points available in the dataset is less than
#'   \code{min.ptos}, then the suboptimal or the supraoptimal models are not
#'   fitted.
#' @param tops method used to divide the dataset in suboptimal and supraoptimal
#'   sections. "Max value" splits the data by the temperature that produces the
#'   highest seed germination rate. "Max R2" splits the data by the temperature
#'   that maximises the R2 of the suboptimal and supraoptimal linear
#'   regressions.
#' @param fractions percentiles into which the seed population is split to fit
#'   the thermal time model. The default is the 9 deciles (i.e. t10, t20.. t90)
#'   as used by Garcia-Huidobro.
#' @return \code{huidobro} returns a S3 object of class "huidobro" with the
#'   results of fitting the thermal time model. The generic functions
#'   \code{summary} and \code{plot} are used to obtain and visualize the model
#'   results.
#' @examples
#' # format dataset with physiodata
#' malva <- physiodata(subset(centaury, population == "La Malva"), x = "temperature")
#' # huidobro() uses the $proportions element within the physiodata object
#' h <- huidobro(malva$proportions)
#' h # prints the main thermal time variables
#' summary(h) # returns the main thermal time variables as a data.table
#' plot(h) # plots the fitted model
#' @references Bewley, J. D., Bradford, K. J., Hilhorst, H. W., & Nonogaki, H.
#'   (2013). Thermal Time Models. In Seeds: Physiology of Development,
#'   Germination and Dormancy, 3rd Edition (pp. 312-317). Springer, New York,
#'   NY. Bradford, K. J. (1990). A water relations analysis of seed germination
#'   rates. Plant Physiology, 94(2), 840-849.
#'
#'   Garcia-Huidobro, J., Monteith, J. L., & Squire, G. R. (1982). Time,
#'   temperature and germination of pearl millet (Pennisetum typhoides S. & H.)
#'   I. Constant temperature. Journal of Experimental Botany, 33(2), 288-296.
#'
#'   Gummerson, R. J. (1986). The effect of constant temperatures and osmotic
#'   potentials on the germination of sugar beet. Journal of Experimental
#'   Botany, 37(6), 729-741.
#' @export
huidobro <- function(d, min.ptos = 3, tops = c("Max R2","Max value"),
                     fractions = (1:9)/10) tryCatch(
                       {
                         rates <- d[, .(fraction = fractions, rate = rates(d = .SD, fractions = fractions)), by = c("treatment")]
                         rates <- rates[! is.na(rates$rate)]

                         if ("Max value" %in% tops)  # Tb<=Tmin, Tc>=Tmax for all quantiles
                         {
                           Ts <- rates[, cardinals(.SD[["treatment"]], rate, which.max(rate),
                                                   min.ptos = min.ptos), by = c("fraction")]
                           Topt <- Ts[, .(Tb = min(c(mean(Tb, na.rm = TRUE), min(Tmin, na.rm = TRUE))),
                                          Tc = max(c(mean(Tc, na.rm = TRUE), max(Tmax), na.rm = TRUE)),
                                          To = mean(To, na.rm = TRUE))]
                         }

                         if ("Max R2" %in% tops)  # Tb<=Tmin, Tc>=Tmax for all quantiles
                         {
                           Ts <- rates[, cardinals(.SD[["treatment"]], rate,
                                                   maximizeR2(.SD[["treatment"]], rate, min.ptos = min.ptos),
                                                   min.ptos = min.ptos), by = c("fraction")]
                           Topt <- Ts[, .(Tb = min(c(mean(Tb, na.rm = TRUE), min(Tmin, na.rm = TRUE))),
                                          Tc =max(c(mean(Tc, na.rm = TRUE), max(Tmax), na.rm = TRUE)),
                                          To = mean(To, na.rm = TRUE))]
                         }

                         d[, probit := qnorm(germination.mean, 0, 1)]

                         subs <- d[treatment <= Topt$To & is.finite(probit)]
                         subs[, thetag := time * (treatment - Topt$Tb)]
                         subfit <- gaussfit(subs)
                         sups <- d[treatment >= Topt$To & is.finite(probit)]
                         sups[, thetag := time * (Topt$Tc - treatment)]
                         supfit <- gaussfit(sups)

                         l <- list()

                         l$data$suboptimal <- subs
                         l$data$supraoptimal <- sups

                         l$parameters$levels$suboptimal <- unique(subs$treatment)
                         l$parameters$levels$supraoptimal <- unique(sups$treatment)
                         l$parameters$cardinals <- Topt
                         l$parameters$suboptimal <- subfit
                         l$parameters$supraoptimal <- supfit

                         class(l) <- "huidobro"
                         l
                       }, error = function(e) e)

# huidobro generic functions

#' @export
print.huidobro <- function(x, ...)
{
  if(is.nan(x$parameters$cardinals$Tc)) {
    cat("Garcia-Huidobro's suboptimal thermal time model", "\n",
        "Suboptimal temperature levels in experiment:",
        round(x$parameters$levels$suboptimal, 1), "\n",
        "Tb - Base temperature:",
        round(x$parameters$cardinals$Tb, 1), "\n",
        "theta50 1 - Suboptimal thermal time (median):",
        round((x$parameters$suboptimal$theta50), 2), "\n",
        "Sigma of the suboptimal thermal time:",
        round((x$parameters$suboptimal$sigma), 2), "\n",
        "R2 of the suboptimal model:",
        round(x$parameters$suboptimal$R2, 2), "\n")} else {
          if(is.nan(x$parameters$cardinals$To)) {
            cat("Garcia-Huidobro's supraoptimal thermal time model", "\n",
                "Supraoptimal temperature levels in experiment:",
                round(x$parameters$levels$supraoptimal, 1), "\n",
                "Tc - Ceiling temperature:",
                round(x$parameters$cardinals$Tc, 1), "\n",
                "theta50 1 - Suboptimal thermal time (median):",
                round(x$parameters$supraoptimal$theta50, 2), "\n",
                "Sigma of the suboptimal thermal time:",
                round(x$parameters$supraoptimal$sigma, 2), "\n",
                "R2 of the suboptimal model:",
                round(x$parameters$supraoptimal$R2, 2), "\n", "\n")} else
                  cat("Garcia-Huidobro's thermal time model", "\n",
                      "Suboptimal temperature levels in experiment:",
                      round(x$parameters$levels$suboptimal, 1), "\n",
                      "Supraoptimal temperature levels in experiment:",
                      round(x$parameters$levels$supraoptimal, 1), "\n",
                      "Tb - Base temperature:",
                      round(x$parameters$cardinals$Tb, 1), "\n",
                      "To - Optimal temperature:",
                      round(x$parameters$cardinals$To, 1), "\n",
                      "Tc - Ceiling temperature:",
                      round(x$parameters$cardinals$Tc, 1), "\n",
                      "theta50 1 - Suboptimal thermal time (median):",
                      round((x$parameters$suboptimal$theta50), 2), "\n",
                      "Sigma of the suboptimal thermal time:",
                      round((x$parameters$suboptimal$sigma), 2), "\n",
                      "R2 of the suboptimal model:",
                      round(x$parameters$suboptimal$R2, 2), "\n",
                      "theta50 2 - Supraoptimal thermal time (median):",
                      round(x$parameters$supraoptimal$theta50, 2), "\n",
                      "Sigma of the supraoptimal thermal time:",
                      round(x$parameters$supraoptimal$sigma, 2), "\n",
                      "R2 of the supraoptimal model:",
                      round(x$parameters$supraoptimal$R2, 2), "\n", "\n")}
}

#' @export
summary.huidobro <- function(object, ...)
{
  data.table(
    method = class(object),
    n.treatments.sub = length(object$parameters$levels$suboptimal),
    n.treatments.sup = length(object$parameters$levels$supraoptimal),
    Tb = object$parameters$cardinals$Tb,
    To = object$parameters$cardinals$To,
    Tc = object$parameters$cardinals$Tc,
    theta50.sub = object$parameters$suboptimal$theta50,
    sigma.sub = object$parameters$suboptimal$sigma,
    R2.sub = object$parameters$suboptimal$R2,
    theta50.sup = object$parameters$supraoptimal$theta50,
    sigma.sup = object$parameters$supraoptimal$sigma,
    R2.sup = object$parameters$supraoptimal$R2)
}

#' @export
plot.huidobro <- function(x, ...) # Plots García-Huidobro's model
{
  if(is.nan(x$parameters$cardinals$Tc)) {
    xpd.status <- par()$xpd
    mar.status <- par()$mar

    par(mar = mar.status + c(0, 0, 0, 4))
    colnumber <- length(unique(x$data$suboptimal$treatment))
    colramp <- colorRampPalette(c("red", "orange", "yellow",
                                  "green", "blue", "violet"))
    x$data$suboptimal$color <- colramp(colnumber)[as.numeric(cut(x$data$suboptimal$treatment, breaks = colnumber))]

    plot(x$data$suboptimal$thetag, x$data$suboptimal$probit, col = x$data$suboptimal$color, pch = 16,
         xlab = expression(paste("Suboptimal ", theta, " (g)")), ylab = "Probit germination")
    abline(lm(x$data$suboptimal$probit ~ x$data$suboptimal$thetag))
    par(xpd = TRUE)
    legend(max(x$data$suboptimal[, thetag]) + max(x$data$suboptimal[, thetag])*.05, 2,
           title = "U+00B0C",
           legend = levels(as.factor(round(x$data$suboptimal$treatment, 1))), pch = 16,
           col = colramp(colnumber))
    par(xpd = xpd.status)
    par(mar = mar.status)} else
    {if(is.nan(x$parameters$cardinals$To)) {
      xpd.status <- par()$xpd
      mar.status <- par()$mar

      par(mar = mar.status + c(0, 0, 0, 4))
      colnumber <- length(unique(x$data$supraoptimal$treatment))
      colramp <- colorRampPalette(c("red", "orange", "yellow",
                                    "green", "blue", "violet"))
      x$data$supraoptimal$color <- colramp(colnumber)[as.numeric(cut(x$data$supraoptimal$treatment, breaks = colnumber))]

      plot(x$data$supraoptimal$thetag, x$data$supraoptimal$probit, col = x$data$supraoptimal$color, pch = 16,
           xlab = expression(paste("Supraoptimal ", theta, " (g)")), ylab = "Probit germination")
      abline(lm(x$data$supraoptimal$probit ~ x$data$supraoptimal$thetag))
      par(xpd = TRUE)
      legend(max(x$data$supraoptimal[, thetag]) + max(x$data$supraoptimal[, thetag])*.05, 2,
             title = "U+00B0C",
             legend = levels(as.factor(round(x$data$supraoptimal$treatment, 1))), pch = 16,
             col = colramp(colnumber))
      par(xpd = xpd.status)
      par(mar = mar.status)} else {
        xpd.status <- par()$xpd
        mar.status <- par()$mar
        ask.status <- par()$ask
        par(ask = TRUE)
        par(mar = mar.status + c(0, 0, 0, 4))

        colnumber <- length(unique(x$data$suboptimal$treatment))
        colramp <- colorRampPalette(c("red", "orange", "yellow",
                                      "green", "blue", "violet"))
        x$data$suboptimal$color <- colramp(colnumber)[as.numeric(cut(x$data$suboptimal$treatment, breaks = colnumber))]

        plot(x$data$suboptimal$thetag, x$data$suboptimal$probit, col = x$data$suboptimal$color, pch = 16,
             xlab = expression(paste("Suboptimal ", theta, " (g)")), ylab = "Probit germination")
        abline(lm(x$data$suboptimal$probit ~ x$data$suboptimal$thetag))
        par(xpd = TRUE)
        legend(max(x$data$suboptimal[, thetag]) + max(x$data$suboptimal[, thetag])*.05, 2,
               title = "U+00B0C",
               legend = levels(as.factor(round(x$data$suboptimal$treatment, 1))), pch = 16,
               col = colramp(colnumber))
        par(xpd = xpd.status)

        colnumber <- length(unique(x$data$supraoptimal$treatment))
        colramp <- colorRampPalette(c("red", "orange", "yellow",
                                      "green", "blue", "violet"))
        x$data$supraoptimal$color <- colramp(colnumber)[as.numeric(cut(x$data$supraoptimal$treatment, breaks = colnumber))]

        plot(x$data$supraoptimal$thetag, x$data$supraoptimal$probit, col = x$data$supraoptimal$color, pch = 16,
             xlab = expression(paste("Supraoptimal ", theta, " (g)")), ylab = "Probit germination")
        abline(lm(x$data$supraoptimal$probit ~ x$data$supraoptimal$thetag))
        par(xpd = TRUE)
        legend(max(x$data$supraoptimal[, thetag]) + max(x$data$supraoptimal[, thetag])*.05, 2,
               title = "U+00B0C",
               legend = levels(as.factor(round(x$data$supraoptimal$treatment, 1))), pch = 16,
               col = colramp(colnumber))

        par(xpd = xpd.status)
        par(mar = mar.status)
        par(ask = ask.status)
      }}
}

# huidobro internal functions

cardinals <- function(x, r, pos, min.ptos = 3)
{
  n <- length(r)
  a1 <- NULL
  b1 <- NULL

  if (pos >= min.ptos)
  {
    yy <- x[1:pos]
    xx <- r[1:pos]
    a1 <- cov(xx, yy) / var(xx)
    b1 <- mean(yy) - a1 * mean(xx)
  }

  a2 <- NULL
  b2 <- NULL

  if ((n - pos + 1) >= min.ptos)
  {
    yy <- x[pos:n]
    xx <- r[pos:n]
    a2 <- cov(xx,yy) / var(xx)
    b2 <- mean(yy) - a2 * mean(xx)
  }

  na <- as.double(NA)

  if (is.null(b1))
  {
    if (is.null(b2)) return(list(Tb = na, Tc = na, To = na, Ro = na,
                                 Tmin = min(x, na.rm = FALSE), Tmax = max(x, na.rm = FALSE),
                                 Nb = pos, Nc = n - pos + 1))
    return(list( Tb = na, Tc = b2, To = a2 * r[pos] + b2, Ro = r[pos],
                 Tmin=min(x, na.rm = FALSE), Tmax = max(x, na.rm = FALSE),
                 Nb = pos, Nc = n - pos + 1))
  } else
  {
    if (is.null(b2)) return(list(Tb = b1, Tc = na, To = a1 * r[pos] + b1,
                                 Ro = r[pos], Tmin = min(x, na.rm = FALSE),
                                 Tmax = max(x, na.rm = FALSE), Nb = pos, Nc = n - pos + 1))
    Ro = (b2 - b1) / (a1 - a2)
    return(list(Tb = b1, Tc = b2, To = a1 * Ro + b1, Ro = Ro,
                Tmin = min(x, na.rm = FALSE), Tmax = max(x, na.rm = FALSE),
                Nb = pos, Nc = n - pos + 1))
  }
}

maximizeR2 <- function(x, r, min.ptos = 3)
{
  n <- length(r)
  if (n < min.ptos) return(NA)
  opt <- as.integer(0)
  opt.val <- 0
  for (pos in min.ptos:(n - min.ptos + 1))
    tryCatch(
      {
        R2 <- 0

        yy <- x[1:pos]
        xx <- r[1:pos]
        cv <- cov(xx, yy)
        if (cv <= 0) next
        R2 <- (cv^2) / (var(xx) * var(yy))

        yy <- x[pos:n]
        xx <- r[pos:n]
        cv <- cov(xx, yy)
        if (cv >= 0) next
        R2 <- R2 + (cv^2) / (var(xx) * var(yy))

        if (R2 > opt.val)
        {
          opt <- pos
          opt.val <- R2
        }
      }, error = function(e) e)

  cv <- cov(x, r)
  R2 <- (cv^2) / (var(x) * var(r))
  if (cv > 0) if (R2 > opt.val) return(n)
  if (cv < 0) if (R2 > opt.val) return(1L)
  return(opt)
}

gaussfit <- function(ddd) tryCatch(
  {
    x <- ddd[["thetag"]]
    probit <- ddd[["probit"]]

    filter <- x > 0
    x <- x[filter]
    probit <- probit[filter]

    hlm <- lm(probit ~ x)
    b <- summary(hlm)$coefficients[1, 1]
    m <- summary(hlm)$coefficients[2, 1]

    theta50 <- -b/m
    sigma <- 1/m

    data.frame(theta50, sigma, R2 = summary(hlm)$r.squared)
  }, error = function(e) e)


R/physiotime.R
#' Fits physiological time seed germination models
#'
#' \code{physiotime} fits physiological time models (thermal time, hydrotime) to
#' seed germination data. It is a wrapper function that transforms data to class
#' "physiodata" and allows to specify the physiological time model to be fitted
#' (i.e. Bradford's hydrotime model or Garcia-Huidobro's thermal time model).
#'
#' @usage physiotime(d, t = "times", g = "germinated", pg = "germinable", x =
#'   "treatment", groups = NULL, method = "bradford", min.ptos = 3, tops =
#'   c("Max R2","Max value"), fractions = (1:9)/10)
#' @param d a data.frame containing the results of a germination experiment. The
#'   data frame should include columns with scoring times, germination counts
#'   (not cumulative), number of potentially germinable seeds, and the
#'   environmental variable of interest. (e.g. temperature or water potential)
#'   (see \code{\link{grasses}} example dataset for appropriate structure).
#' @param t the name of a column in \code{d} containing a vector of numeric
#'   scoring times.
#' @param g the name of a column in \code{d} containing a vector of integer
#'   germination counts (non cumulative).
#' @param pg the name of a column in \code{d} containing a vector of integer
#'   numbers of potentially germinable seeds.
#' @param x the name of a column in \code{d} containing a vector of numeric
#'   values for the environmental variable of interest (e.g. temperature, water
#'   potential).
#' @param groups optional, the names of columns in \code{d} containing grouping
#'   variables for the experiment that have to be analysed separately (e.g.
#'   different species or populations, different temperatures in a water
#'   potential experiment, different treatments to break seed dormancy).
#' @param method the method to be used to fit the models, can be "bradford" to
#'   fit a hydrotime model or "huidobro" to fit a thermal time model.
#' @param min.ptos minimal number of data points (i.e. different temperature
#'   treatments) needed to fit the suboptimal and supraoptimal germination
#'   models if fitting a thermal time model. If the number of points available
#'   in the dataset is less than \code{min.ptos}, then the suboptimal or the
#'   supraoptimal models are not fitted.
#' @param tops method used to divide the dataset in suboptimal and supraoptimal
#'   sections if fitting a thermal time model. "Max value" splits the data by
#'   the temperature that produces the highest seed germination rate. "Max R2"
#'   splits the data by the temperature that maximises the R2 of the suboptimal
#'   and supraoptimal linear regressions.
#' @param fractions percentiles into which the seed population is split if
#'   fitting a thermal time model. The default is the 9 deciles (i.e. t10, t20..
#'   t90) as used by Garcia-Huidobro.
#' @return \code{physiotime} returns a S3 object of class "physiotime".
#'  The object is a list containing, for each group (seedlot, species, etc.) the
#'  results of fitting the physiological time models.  The generic functions
#'   \code{summary} and \code{plot} are used to obtain and visualize the model
#'   results.
#' @examples
#' m <- physiotime(centaury, x = "temperature",
#'                 method = "huidobro", groups = c("species", "population"))
#' m
#' summary(m)
#' plot(m)
#' @export
physiotime <- function(d, t = "times", g = "germinated", pg = "germinable",
                       x = "treatment", groups = NULL, method = "bradford",
                       min.ptos = 3, tops = c("Max R2","Max value"),
                       fractions = (1:9)/10)
{
  d <- physiodata(d, t, g, pg, x, groups)
  if(is.null(groups)) {
    z <- get(method)(d$proportions)
  } else {
    listd <- split(d$proportions, by = groups, drop = TRUE)
    z <- lapply(listd, function(d) {
      zi <- get(method)(d)
      zi$groups <- unique(d[, ..groups])
      zi$groups <- zi$groups[, group := do.call(paste, Map(function(x, y)
        paste(x, y, sep = ": "), names(.SD), .SD))]
      zi
    })
    class(z) <- "physiotime"}
  z
}

# physiotime generic functions

#' @export
print.physiotime <- function(x, ...)
{
  cat("A list of physiological time germination models", "\n",
      "calculated for the following", length(x), "groups:", "\n", "\n")
  for(y in x) {
    cat(y$groups$group, "\n")
    print(y)
  }
}

#' @export
summary.physiotime <- function(object, ...)
{
  l <- lapply(object, function(y) data.frame(
    y$groups[,-ncol(y$groups), with = FALSE], summary(y)))
  rbindlist(l)
}

#' @export
plot.physiotime <- function(x, ...) # Plots Bradford's model
{
  ask.status <- par()$ask
  par(ask = TRUE)
  for(y in x) {
    plot(y)
    title(main = y$groups$group)
  }
  par(ask = ask.status)
}


R/bradford.R
#' Fits Bradford's hydrotime model
#'
#' \code{bradford} fits a hydrotime seed germination model using the method of
#' Bradford (Gummerson 1986, Bradford 1990, Bewley et al. 2013). This function
#' can be used only with one-group dataset, i.e. one seed lot of one species. To
#' fit models to grouped datasets (multi-seedlots, multi-species) use the
#' function \code{physiotime} instead.
#'
#' @usage bradford(d)
#' @param d a data.table within a "physiodata" object, containing the cumulative
#'   germination proportion at each scoring time and water potential treatment.
#' @return \code{bradford} returns a S3 object of class "bradford" with the
#'   results of fitting the hydrotime model. The generic functions
#'   \code{summary} and \code{plot} are used to obtain and visualize the model
#'   results.
#' @examples
#' # format dataset with physiodata
#' anisantha <- physiodata(subset(grasses, species == "Anisantha rubens"), x = "psi")
#' # bradford() uses the $proportions element within the physiodata object
#' b <- bradford(anisantha$proportions)
#' b # prints the main hydrotime variables
#' summary(b) # returns the main hydrotime variables as a data.table
#' plot(b) # plots the fitted model
#' @references Bewley, J. D., Bradford, K. J., Hilhorst, H. W., & Nonogaki, H.
#'   (2013). Hydrotime Model of Germination. In Seeds: Physiology of
#'   Development, Germination and Dormancy, 3rd Edition (pp. 303-307). Springer,
#'   New York, NY.
#'
#'   Bradford, K. J. (1990). A water relations analysis of seed germination
#'   rates. Plant Physiology, 94(2), 840-849.
#'
#'   Gummerson, R. J. (1986). The effect of constant temperatures and osmotic
#'   potentials on the germination of sugar beet. Journal of Experimental
#'   Botany, 37(6), 729-741.
#' @export
bradford <- function(d) tryCatch(
  {
    d <- d[! germination.mean %in% c(0, 1)]
    d[, probit := qnorm(germination.mean, 0, 1)]
    theta <- bradtheta(d[, probit], d[, treatment], 1/d[, time])
    d[, psibg := treatment - theta / time]

    hlm <- lm(probit ~ psibg, data = d)
    b <- summary(hlm)$coefficients[1, 1]
    m <- summary(hlm)$coefficients[2, 1]

    z <- list()

    z$data <- d

    z$parameters$levels <- d[, .(Levels = unique(treatment))][, Levels]
    z$parameters$theta <- theta
    z$parameters$psib50 <- -b/m
    z$parameters$sigma <- 1/m
    z$parameters$R2 <- summary(hlm)$r.squared

    class(z) <- "bradford"
    z
  }, error = function(e) e)

# bradford generic functions

#' @export
print.bradford <- function(x, ...)
{
  cat("Bradford's hydrotime model", "\n",
      "Water potential levels in experiment:",
      round(x$parameters$levels, 1), "\n",
      "Theta - Hydrotime constant:",
      round(x$parameters$theta, 2), "\n",
      "Psib50 - Base water potential (median):",
      round((x$parameters$psib50), 2), "\n",
      "Sigma of the base water potential:",
      round(x$parameters$sigma, 2), "\n",
      "R2:", round(x$parameters$R2, 2), "\n", "\n")
}

#' @export
summary.bradford <- function(object, ...)
{
  data.table(
    method = class(object),
    n.treatments = length(object$parameters$levels),
    theta = object$parameters$theta,
    psib50 = object$parameters$psib50,
    sigma = object$parameters$sigma,
    R2 = object$parameters$R2)
}

#' @export
plot.bradford <- function(x, ...) # Plots Bradford's model
{
  colnumber <- length(unique(x$data$treatment))
  colramp <- colorRampPalette(c("red", "orange", "yellow",
                                "green", "blue", "violet"))
  x$data$color <- colramp(colnumber)[as.numeric(cut(x$data$treatment, breaks = colnumber))]

  plot(x$data$psibg, x$data$probit, col = x$data$color, pch = 16,
       xlab = expression(paste(psi[b], " (g)")), ylab = "Probit germination")
  abline(lm(x$data$probit ~ x$data$psibg))
  legend("topleft", title = expression(paste(psi, " (MPa)")),
         legend = levels(as.factor(round(x$data$treatment, 1))), pch = 16,
         col = colramp(colnumber))
}

# bradford internal functions

bradtheta <- function(probit, x, invt) # Calculates Bradford's Theta
{
  finite <- is.finite(probit) & is.finite(x) & is.finite(invt)
  probit <- probit[finite]
  x <- x[finite]
  invt <- invt[finite]

  Sxy <- cov(probit, x)
  Sxz <- cov(probit, invt)
  Sxx <- var(probit)
  Syz <- cov(x, invt)
  Syy <- var(x)
  Szz <- var(invt)

  a <- -2 * (Sxz ^ 2) * Syz + 2 * Sxy * Sxz * Szz
  b <- 2 * (Sxz ^ 2) * Syy + 4 * Sxy * Syz * Sxz - 2 * (Sxy ^ 2) *
    Szz - 4 * Syz * Sxy * Sxz
  c <- 2 * Syz * Sxy ^ 2 - 2 * Sxy * Sxz * Syy

  max(c(( -b + sqrt(b ^ 2 - 4 * a * c))/(2 * a),
        (-b -sqrt(b ^ 2 - 4 * a * c))/(2 * a)))
}
